docs/source/migration.mdx

```diff
@@ -545,6 +566,52 @@ await startStandaloneServer(server, {
 
 </MultiCodeBlock>
 
+If you want to migrate quickly to Apollo Server 4 *without* altering your data sources, the snippet below replicates Apollo Server 3's `dataSources` behavior with a custom plugin.
+
+<ExpansionPanel title="Click to expand">
+
+```ts
+import { ApolloServerPlugin, BaseContext } from '@apollo/server';
+import type { DataSource } from 'apollo-datasource';
+
+type DataSources = Record<string, DataSource>;
+type DataSourcesFn<TDataSources extends DataSources = DataSources> =
+  () => TDataSources;
+
+interface ContextWithDataSources<TDataSources extends DataSources = DataSources>
+  extends BaseContext {
+  dataSources?: TDataSources;
+}
+
+export const ApolloDataSources = <
+  TDataSources extends DataSources = DataSources,
+>(options: {
+  dataSources: DataSourcesFn<TDataSources>;
+}): ApolloServerPlugin<ContextWithDataSources<TDataSources>> => ({
```

Comment:
I'm not sure what problem the `TDataSources` generic is solving - seems like hard-typing that `Record<string, DataSource>` everywhere is sufficient, and I'm not sure why (or even how?) I might extend that if I was using this snippet.

Reply:
I'm not too sure about the answer here, @smyrick would you happen to know?

Reply:
it's required for the DataSourcesFn, otherwise `options.dataSources()` returns an `Any` type.

Reply:
I see the usefulness of the `DataSourcesFn`, but I guess I don't see why it can't just be:

```ts
type DataSourcesFn =  () => DataSources;
```

or in other words, as the consumer of these types, why would I ever need to plug in my own type for `TDataSources` rather than rely on the default which is the `DataSources` record?

Reply:
Sure I agree, I don't think there is really any added benefit to the optional type parameter.

@rkoron007 could probably simplify the whole example to just...
```
import { ApolloServerPlugin, BaseContext } from '@apollo/server';
import type { DataSource } from 'apollo-datasource';

type DataSources = Record<string, DataSource>;
type DataSourcesFn = () => DataSources;

interface ContextWithDataSources extends BaseContext {
  dataSources?: DataSources
}

export const ApolloDataSources = (options: {dataSources: DataSourcesFn}):ApolloServerPlugin<ContextWithDataSources> => ({
  requestDidStart: async (requestContext) => {
    const dataSources = options.dataSources();
    const initializers = Object.values(dataSources).map(async (dataSource) => {
      if (dataSource.initialize) dataSource.initialize({cache: requestContext.cache, context: requestContext.contextValue})
    });

    await Promise.all(initializers);  
    
    requestContext.contextValue.dataSources = dataSources;
  }
});
```

Reply:
Okay, I'll do that! Thank you both so much 🚀